Non concurrent mode now sends connect network response as per spec

simonhmorris1 opened the pull request:
Fixes #31660, Implement Non-Concurrent Spec Change #8728 in the SDK.

This change now sends the ConnectNetworkResponse(Success) on the commissioning channel before closing it and enabling the operation channel as required by the https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/8728 spec change.

This changes the commissioner and Linux Raspberry Pi lighting app commissionee. The Non-concurrent mode in lighting app requires examples/lighting-app/linux/include/CHIPProjectAppConfig.h to be modified as follows to disable concurrent connection mode. `#define CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION 0`

This has been tested with the `CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION` set to zero and one, the passing test logs are attached.

[chip_output_concurrent_mode.txt](https://github.com/project-chip/connectedhomeip/files/14082223/chip_output_concurrent_mode.txt)
[light-app-concurrent.txt](https://github.com/project-chip/connectedhomeip/files/14082224/light-app-concurrent.txt)
[light-app-non-concurrent.txt](https://github.com/project-chip/connectedhomeip/files/14082225/light-app-non-concurrent.txt)
[chip_output_non-concurrent_mode.txt](https://github.com/project-chip/connectedhomeip/files/14082227/chip_output_non-concurrent_mode.txt)

Damian-Nordic commented on src/platform/Linux/BLEManagerImpl.cpp:
```diff
@@ -215,12 +219,29 @@ void BLEManagerImpl::_OnPlatformEvent(const ChipDeviceEvent * event)
+#if !CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION
+        if (GetBleTerminating())
+        {
+            ChipLogProgress(DeviceLayer, "Non-concurrent mode BLE packet acknowledged. Sending CloseAllBLEConnections");
```
Isn't this assuming that one indication will be enough to transfer the entire ConnectNetworkResponse (no fragmentation)? Maybe this should rather be triggered somehow from inside `BLEEndpoint` code that checks if `(mBtpEngine.TxState() == BtpEngine::kState_Complete)`.

simonhmorris1 replied:
Updated to use BTP.

Damian-Nordic commented on src/ble/BleLayer.cpp:
```diff
@@ -789,6 +790,18 @@ void BleLayer::OnConnectionError(void * appState, CHIP_ERROR err)
 }
 
+// BLE Flag to wait for packet transmission confirmation
+static bool BleTerminating = false;
```
Why using a global if `BleLayer` is a class that can have regular members? Also, how about a name such as `mTerminateOnPacketTxComplete` or something that better describes the behavior?

simonhmorris1 replied:
Changed to use regular member.

adabreuti commented:
@simonhmorris1 Great catch and change, one note is that these changes aren't currently functional for a for a Thread based platform implementation, as there is no Thread based event to call HandleConcurrentConnectNetwork().
https://github.com/simonhmorris1/connectedhomeip/blob/master/src/app/clusters/network-commissioning/network-commissioning.cpp#L1066

 I'm wondering if it would be useful for other vendors if we added an additional API to the DeviceControlServer for Thread that would simply post a kThreadDeviceAvailable event for vendors to implement in their respective BLEManager's since this is currently in a shared file.

This would allow for non-concurrent operation on Thread following a similar path as to what you have implemented for Wi-Fi.
https://github.com/simonhmorris1/connectedhomeip/blob/master/src/platform/DeviceControlServer.cpp#L83

Let me know your thoughts.

simonhmorris1 commented:
> @simonhmorris1 Great catch and change, one note is that these changes aren't currently functional for a for a Thread based platform implementation, as there is no Thread based event to call HandleConcurrentConnectNetwork(). https://github.com/simonhmorris1/connectedhomeip/blob/master/src/app/clusters/network-commissioning/network-commissioning.cpp#L1066
> 
> I'm wondering if it would be useful for other vendors if we added an additional API to the DeviceControlServer for Thread that would simply post a kThreadDeviceAvailable event for vendors to implement in their respective BLEManager's since this is currently in a shared file.
> 
> This would allow for non-concurrent operation on Thread following a similar path as to what you have implemented for Wi-Fi. https://github.com/simonhmorris1/connectedhomeip/blob/master/src/platform/DeviceControlServer.cpp#L83
> 
> Let me know your thoughts.



> @simonhmorris1 Great catch and change, one note is that these changes aren't currently functional for a for a Thread based platform implementation, as there is no Thread based event to call HandleConcurrentConnectNetwork(). https://github.com/simonhmorris1/connectedhomeip/blob/master/src/app/clusters/network-commissioning/network-commissioning.cpp#L1066
> 
> I'm wondering if it would be useful for other vendors if we added an additional API to the DeviceControlServer for Thread that would simply post a kThreadDeviceAvailable event for vendors to implement in their respective BLEManager's since this is currently in a shared file.
> 
> This would allow for non-concurrent operation on Thread following a similar path as to what you have implemented for Wi-Fi. https://github.com/simonhmorris1/connectedhomeip/blob/master/src/platform/DeviceControlServer.cpp#L83
> 
> Let me know your thoughts.

Yep I'll take a look at making that change.

bzbarsky-apple commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -135,6 +135,24 @@ void Instance::Shutdown()
     mpBaseDriver->Shutdown();
 }
 
+#if !CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION
```
Is CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION defined for devices that don't do BLE at all?  I guess probably....

simonhmorris1 replied:
Yes, the default is is defined in src/include/platform/CHIPDeviceConfig.h

bzbarsky-apple replied:
No, what I meant is: is it going to be true for non-BLE devices?  But I am guessing yes.

bzbarsky-apple commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -135,6 +135,24 @@ void Instance::Shutdown()
+        return;
+    }
+
+    chip::DeviceLayer::ConnectivityMgr().GetBleLayer()->mTerminateOnPacketTxComplete = true;
```
```suggestion
    DeviceLayer::ConnectivityMgr().GetBleLayer()->mTerminateOnPacketTxComplete = true;
```
This is all inside `namespace chip`.

simonhmorris1 replied:
Done.

bzbarsky-apple commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -135,6 +135,24 @@ void Instance::Shutdown()
+        return;
+    }
+
+    chip::DeviceLayer::ConnectivityMgr().GetBleLayer()->mTerminateOnPacketTxComplete = true;
```
Please include the right headers for these bits.

simonhmorris1 replied:
Added `#include <platform/ConnectivityManager.h>`

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -3062,28 +3041,9 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
-                return;
-            }
-        }
+                        params.GetSupportsConcurrentConnection().Value());
```
So now we are reading the "supports concurrent connection" bit just so we can log it here?

simonhmorris1 replied:
Yes, with the change to always send ConnectNetworkResponse(Success) on BLE prior to closing the commissioning channel and starting the operational channel. However I can see additional optimisation that could be done by the controller if knowing that this is a non-concurrent device.

bzbarsky-apple commented on src/platform/Linux/BLEManagerImpl.cpp:
Other platforms need similar fixes, right?  Presumably tracked in some set of issues?

simonhmorris1 replied:
If other platforms want to support Non-Concurrent mode then yes similar changes would be needed, however they would only use non-concurrent mode if the platform does not support concurrent mode. I have used the Raspberry Pi as the example as that is used in QA and certification testing, so it can now do either. I don't think any tickets need raising on the other platforms as they all support concurrent mode. New platforms that do not support Concurrent mode will need these changes.

simonhmorris1 commented on src/ble/BleApplicationDelegate.h:
```diff
@@ -42,6 +42,10 @@ class DLL_EXPORT BleApplicationDelegate
     virtual void NotifyChipConnectionClosed(BLE_CONNECTION_OBJECT connObj) = 0;
+
+    // Called to determine whether the BLE connection should be closed when in Non-concurrent mode if sending
+    // ConnectNetworkResponse. The BTP will be in kState_Complete o all fragments have been sent.
```
Done.

simonhmorris1 commented on src/ble/BleLayer.h:
```diff
@@ -233,6 +233,9 @@ class DLL_EXPORT BleLayer
     BleLayerDelegate * mBleTransport = nullptr;
 
+    // Set in Non-concurrent mode when the ConnectNetworkResponse has been requested
+    bool mTerminateOnPacketTxComplete = false;
```
Now use mState instead of mTerminateOnPacketTxComplete.

tcarmelveilleux commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -135,6 +136,24 @@ void Instance::Shutdown()
+        return;
+    }
+
+    DeviceLayer::ConnectivityMgr().GetBleLayer()->mState = Ble::BleLayer::kState_Disconnecting;
```
Why are we directly touching BLE here? We should NOT be touching internal state of BLE layer from this place, and at the very least, if done, it should not be via member fields.

simonhmorris1 replied:
I looked at sending an event to the Commissioning Manager which would call the BLE, but that has race conditions. The disconnecting state needs setting prior to the sending of the BTP packet containing the ConnectNetworkResponse(Success) but the event is queued at the same time AddResponse() is called. Which means there is the potential that the BTP packet completes prior to the event being handled. Therefore I have created a function which is now called to set the mState within the BLE layer rather than directly accessing the member field.

tcarmelveilleux commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -135,6 +136,24 @@ void Instance::Shutdown()
+        return;
+    }
+
+    DeviceLayer::ConnectivityMgr().GetBleLayer()->mState = Ble::BleLayer::kState_Disconnecting;
```
This needs to be conditional on CONFIG_NETWORK_LAYER_BLE

simonhmorris1 replied:
Done

tcarmelveilleux commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -177,12 +196,7 @@ void Instance::InvokeCommand(HandlerContext & ctxt)
-        // If commissionee does not support Concurrent Connections, request the BLE to be stopped.
-        // Start the ConnectNetwork, but this will not complete until the BLE is off.
-        ChipLogProgress(NetworkProvisioning, "Closing BLE connections due to non-concurrent mode");
-        DeviceLayer::DeviceControlServer::DeviceControlSvr().PostCloseAllBLEConnectionsToOperationalNetworkEvent();
```
Can this leave some existing devices in a bad state?

simonhmorris1 replied:
I do not think so, the CloseAllBleConnections is now done on the confirmation that the BTP packet containing the ConnectNetworkResponse(Success) has been sent.

tcarmelveilleux commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -826,13 +844,19 @@ void Instance::HandleQueryIdentity(HandlerContext & ctx, const Commands::QueryId
 void Instance::OnResult(Status commissioningError, CharSpan debugText, int32_t interfaceStatus)
 {
     auto commandHandleRef = std::move(mAsyncCommandHandle);
-    auto commandHandle    = commandHandleRef.Get();
+
+    // In Non-concurrent mode the commandHandle will be null here, the ConnectNetworkResponse
+    // has already been sent and the BLE will have been stopped, however the other functionality
+    // is still required
+#if CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION
+    auto commandHandle = commandHandleRef.Get();
```
What happens to the lifecycle of the commandHandleRef in this case?

simonhmorris1 replied:
The commandHandleRef is completed when the NetworkConnectResponse(Success) is sent in SendNonConcurrentConnectNetworkResponse() prior to getting here.

tcarmelveilleux commented on src/app/clusters/network-commissioning/network-commissioning.h:
```diff
@@ -48,6 +48,10 @@ class Instance : public CommandHandlerInterface,
     void Shutdown();
 
+#if !CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION
+    void SendNonConcurrentConnectNetworkResponse();
```
Why is this public?

simonhmorris1 replied:
Moved

tcarmelveilleux commented on src/app/server/CommissioningWindowManager.cpp:
```diff
@@ -89,8 +89,8 @@ void CommissioningWindowManager::OnPlatformEvent(const DeviceLayer::ChipDeviceEv
     {
-        ChipLogProgress(AppServer, "Received kCloseAllBleConnections");
-        mServer->GetBleLayerObject()->CloseAllBleConnections();
+        ChipLogProgress(AppServer, "Received kCloseAllBleConnections:%d", event->Type);
```
Please `static_cast<int>(event->Type)`

simonhmorris1 replied:
Done

tcarmelveilleux commented on src/app/server/CommissioningWindowManager.cpp:
```diff
@@ -89,8 +89,8 @@ void CommissioningWindowManager::OnPlatformEvent(const DeviceLayer::ChipDeviceEv
-        ChipLogProgress(AppServer, "Received kCloseAllBleConnections");
-        mServer->GetBleLayerObject()->CloseAllBleConnections();
+        ChipLogProgress(AppServer, "Received kCloseAllBleConnections:%d", event->Type);
+        mServer->GetBleLayerObject()->Shutdown();
```
Shutdown != to CloseAllBleConnections. How widely was this change tested? It impacts all BLE users.

simonhmorris1 replied:
As we are in Non-concurrent mode, the BLE must be stopped for all BLE users. The BLE will be either powered off or just the use of the RF disabled. as we are commissioning I am not aware of any other BLE users at this time.

tcarmelveilleux commented on src/include/platform/CHIPDeviceEvent.h:
```diff
@@ -171,7 +171,7 @@ enum PublicEventTypes
      * the BLE device is closed and the WiFi device has been started.
      */
-    kWiFiDeviceAvailable,
+    kOperationalNetworkStarted,
```
This is a change of public API. Changing this may break platforms. This was the wrong name, but the old name must be maintained, and kWiFiDeviceAvailable should just steer to the same place as kOperationalNetworkStarted.

simonhmorris1 replied:
Done, however as this was only added a few weeks ago after the 1.2 release, I do not think it has been used by anyone, so we may not need to do it.

tcarmelveilleux commented on src/platform/Linux/BLEManagerImpl.cpp:
```diff
@@ -222,8 +226,18 @@ void BLEManagerImpl::_OnPlatformEvent(const ChipDeviceEvent * event)
     case DeviceEventType::kCHIPoBLEWriteReceived:
-        HandleWriteReceived(event->CHIPoBLEWriteReceived.ConId, &CHIP_BLE_SVC_ID, &ChipUUID_CHIPoBLEChar_RX,
-                            PacketBufferHandle::Adopt(event->CHIPoBLEWriteReceived.Data));
+#if !CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION
+        // If BLE is being stopped, ignore any write messages
+        if (mState != kState_Initialized)
+        {
+            ChipLogProgress(DeviceLayer, "Non-concurrent mode BLE stopped, BLE write ignored");
+        }
+        else
+#endif
```
It's really not clear why this is being done. Please document more detauks about how we could get here. Why is it specific to non-concurrent? This seems like always an error that could be an early return that is not conditional to the `#if !CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION`

simonhmorris1 replied:
BlueZ sends another message after the BLE has been stopped which was causing the Wi-Fi to be disconnected and reestablished slowing down commissioning. This is a good point, I have modified the system so that BlueZ is disconnected at the same time of the underlying BLE, negating the need for this. So removed.

tcarmelveilleux reviewed (changes requested):
- [ ] Return all existing public API and layer new features in addition to old names, even if they go to the same place
- [ ] Refactor direct access to `DeviceLayer::ConnectivityMgr().GetBleLayer()->mState`
- [ ] This seems risky/brittle given the breadth of changes. Please add an integration test that does not rely on multiple manual tests people have to keep running to prove the functioning of this feature.
- [ ] Make BLEManagerImpl.cpp error handling non-build-conditional wherever they apply uniformly
- [ ] Justify the move to `mServer->GetBleLayerObject()->Shutdown();` and how it is possible to guarantee that after a mid-commissioning failure, the window could still be open for the remainder of the time, to allow more BLE connections to support finishing the commissioning without reboot of the device.

bzbarsky-apple reviewed (approved):
Approving, assuming @tcarmelveilleux's issues are addressed.

simonhmorris1 commented on src/app/clusters/network-commissioning/network-commissioning.cpp:
```diff
@@ -826,13 +844,19 @@ void Instance::HandleQueryIdentity(HandlerContext & ctx, const Commands::QueryId
         // We may receive the callback after it and should make it noop.
         return;
     }
+#endif
```
Done

simonhmorris1 commented on src/include/platform/DeviceControlServer.h:
```diff
@@ -37,7 +37,7 @@ class DeviceControlServer final
     CHIP_ERROR PostConnectedToOperationalNetworkEvent(ByteSpan networkID);
     CHIP_ERROR PostCloseAllBLEConnectionsToOperationalNetworkEvent();
-    CHIP_ERROR PostWiFiDeviceAvailableNetworkEvent();
+    CHIP_ERROR PostOperationalNetworkStartedEvent();
```
Done, however I do not think this has been used by anyone, see other comment.

simonhmorris1 commented:
> * [ ]  Return all existing public API and layer new features in addition to old names, even if they go to the same place
**Done**
> * [ ]  Refactor direct access to `DeviceLayer::ConnectivityMgr().GetBleLayer()->mState`
**Thinking on this one**
> * [ ]  This seems risky/brittle given the breadth of changes. Please add an integration test that does not rely on multiple manual tests people have to keep running to prove the functioning of this feature.
**Please could you point me at instructions of how to add/write integration tests. Could this be done under a different ticket?**
> * [ ]  Make BLEManagerImpl.cpp error handling non-build-conditional wherever they apply uniformly
**Done**
> * [ ]  Justify the move to `mServer->GetBleLayerObject()->Shutdown();` and how it is possible to guarantee that after a mid-commissioning failure, the window could still be open for the remainder of the time, to allow more BLE connections to support finishing the commissioning without reboot of the device.
**See comment about Shutdown in code review also,** 
In non-concurrent mode, the BLE could be powered off, to allow the WiFi to start, think of a scenario of a cut down BLE/Wi-Fi IoT chip in use with limited RAM so only one can be used at a time. Dealing with this is implementation dependent. The code here is a RPi example to ensure the spec signalling is in place. Getting BlueZ working well for production would require more work.
